zerver/lib/user_groups.py

```diff
@@ -45,17 +50,38 @@ def access_user_group_by_id(
     return user_group
 
 
-def access_user_groups_as_potential_subgroups(
+def get_and_lock_user_groups_and_indirect_subgroups(
     user_group_ids: Sequence[int], acting_user: UserProfile
-) -> List[UserGroup]:
-    user_groups = UserGroup.objects.filter(id__in=user_group_ids, realm=acting_user.realm)
+) -> Tuple[List[UserGroup], List[UserGroup]]:
+    """This locks the user groups with the given user_group_ids, as well as their indirect subgroups,
+    and return the queried user group with their indirect subgroups in a tuple.
+    Regardless of whether the user groups returned are used, always call this helper before making
+    changes to subgroup membership. This avoids introducing cycles among user groups when there is a
+    race condition in which one of these subgroups become an ancestor of the parent user group in
+    another transaction.
+
+    Note that it does not do the permission check in access_user_group_by_id.
+    """
+    # TODO: These select_for_update queries are subject to the deadlock issue, and better error handling is needed.
+    # We may use select_for_update(nowait=True) and release the locks held by ending the transaction with a JsonableError
+    # by handling the DatabaseError. But at the current scale of concurrent requests, we rely on Postgres's deadlock
+    # detection and show the users the "Internal Server Error" when it occurs encouraging retries.
+
+    # Calling list with the QuerySet forces its evaluation putting a lock on the queried rows.
+    user_groups = list(
+        UserGroup.objects.select_for_update().filter(id__in=user_group_ids, realm=acting_user.realm)
+    )
+    recursive_subgroups = list(
+        get_recursive_subgroups_for_groups(user_group_ids).select_for_update()
+    )
```

Comment:
> there is a clear ordering (which may be easier to implement in the recursive CTE query), as opposed to the potentially many-branching (and self-overlapping) subgroup query

I'm not sure if I understand the argument about the ordering of subgroups :)

To my knowledge, we can have overlapping supergroups while not violating the DAG constraint, unless I'm missing something:

```
H     A    F    supergroups ↑
 \   / \  /                 
  \ /   B
   C   / \
    \ /   \
     D     E
      \   /                 
        G       subgroups   ↓
```

Should I infer that overlaps are less likely because it is more usual to have 1 supergroup vs. multiple subgroups than having 1 subgroup vs. multiple supergroups, which makes tracing upwards have a clearer ordering?

Reply:
No, you're right that the DAG doesn't force the constraints that for some reason I was assuming it did -- there may be _more likely_ properties, but I don't know that my guesses about which those are would be correct, and I don't think that's a strong enough property to argue for one direction over the other.

We do need to ensure some consistent ordering across all of the locks, regardless of if we go up or down.

Reply:
I agree. So currently our endpoints support the bulk creation of subgroups. A logical ordering would be locking the subgroups to be added, the recursive subgroups of these subgroups, and finally the supergroup. The order should be consistent as long as the recursive query is reused.

One thing that I want to change here though, is to remove the following:
```python
UserGroup.objects.select_for_update().filter(id__in=user_group_ids, realm=acting_user.realm)
```
Because the recursive query also returns the queried user groups themselves in addition to their recursive subgroups.